Add "Alphabetize choices" button for "List of options" type custom profile fields.

tnmkr opened the pull request:
<!-- Describe your pull request here.-->

**Commit 1** Fixes a bug reported in [CZO](https://chat.zulip.org/#narrow/stream/9-issues/topic/Bad.20state.20when.20editing.20options.20for.20custom.20profile.20field.2E/near/1808130).

**Commit 2** Fixes: #28607 
[CZO #design](https://chat.zulip.org/#narrow/stream/137-feedback/topic/Favorite.20editor/near/1836053)

**Changes from previous PR #28867** 
- We disable the button instead of hiding it, this was decided in [CZO thread](https://chat.zulip.org/#narrow/stream/137-feedback/topic/Favorite.20editor/near/1721247) linked in the issue.
- I do not fully understand the approach used in that PR and I think it's better to build this feature around `SortableJS.sort()`, so I started from scratch in this PR.

**Screenshots and screen captures:**

| Before | After |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/133781250/8b85ff8a-e35f-47b4-81a9-4d97d1417459) | ![image](https://github.com/zulip/zulip/assets/133781250/3447b98a-2296-47a6-8f9a-9b565558a522) |
| ![image](https://github.com/zulip/zulip/assets/133781250/1bc8b2d6-9e62-4802-b1a5-04ea0a7482b8) | ![image](https://github.com/zulip/zulip/assets/133781250/0ce25d7f-5c05-4f7e-bde8-42bb58cce2d6) | 

| Enabled | Disabled |
|--------|--------|
| ![image](https://github.com/zulip/zulip/assets/133781250/a52522a6-7a5c-44db-a87b-4020a8c29af5) | ![image](https://github.com/zulip/zulip/assets/133781250/2d908231-2f31-44a9-a0be-a1e1bdb4844d) | 
| ![image](https://github.com/zulip/zulip/assets/133781250/e3d0cae6-a4fc-4de4-9707-0ba72557f765) | ![image](https://github.com/zulip/zulip/assets/133781250/37e9bc31-e4cc-4408-8a7c-1623c1243997) |

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

tnmkr commented on web/styles/settings.css:
```diff
@@ -1627,6 +1627,11 @@ $option_title_width: 180px;
 
+.profile-field-choices-wrapper > .alphabetize-choices-button {
+    display: block;
+    margin: 12px 0 0;
```
Got this value from previous PR #28867.

tnmkr commented:
@sanchi-t This PR is ready for review. Please take a look when possible :)

@zulipbot add "buddy review"

sanchi-t commented on web/templates/settings/add_new_custom_profile_field_form.hbs:
```diff
@@ -26,11 +26,12 @@
             <table class="profile_field_choices_table">
                 <tbody id="profile_field_choices" class="profile-field-choices"></tbody>
             </table>
+            <button disabled type="button" class="button white rounded alphabetize-choices-button">{{t "Alphabetize choices" }}</button>
```
Why disable the button when `maybe_toggle_alphabetize_button` is called upon clicking the edit button of the profile field?


Also, the commit message should discuss changes in the commit but you maybe accidentally write "PR"?

sanchi-t replied:
Oh okay got it.

sanchi-t commented on web/src/settings_profile_fields.ts:
```diff
@@ -752,10 +843,26 @@ function set_up_select_field(): void {
+    );
     $profile_field_choices.on("click", "button.delete-choice", function (this: HTMLElement) {
         delete_choice_row(this);
+        maybe_toggle_alphabetize_button($profile_field_choices);
```
If the button is already disabled, indicating that the fields are sorted, then we should not call this function when deleting a choice, as deleting a choice from sorted fields will still keep them sorted. Right?

tnmkr replied:
Yep, you are right! Thanks for finding this optimization!

I have added a parameter and the actual check inside `maybe_toggle_alphabetize_button` function. This is because we already have a reference to the button in `maybe_toggle_alphabetize_button`.

```diff
diff --git a/web/src/settings_profile_fields.ts b/web/src/settings_profile_fields.ts
index 12cbad1746..9f2d2e62fa 100644
--- a/web/src/settings_profile_fields.ts
+++ b/web/src/settings_profile_fields.ts
@@ -440,7 +440,19 @@ function alphabetize_profile_field_choices(
     $("#edit-custom-profile-field-form-modal .dialog_submit_button").prop("disabled", false);
 }
 
-function maybe_toggle_alphabetize_button($sortable_element: JQuery): void {
+function maybe_toggle_alphabetize_button(
+    $sortable_element: JQuery,
+    action_unsorts_field = true,
+): void {
+    // This is safe because only one profile field modal is open at a time ever.
+    const $alphabetize_choices_button = $(
+        ".profile-field-choices-wrapper > button.alphabetize-choices-button",
+    );
+
+    if (!action_unsorts_field && $alphabetize_choices_button.prop("disabled")) {
+        // The fields are already sorted.
+        return;
+    }
     assert($sortable_element[0] !== undefined);
     const sortable_instance = SortableJS.get($sortable_element[0]);
     assert(sortable_instance !== undefined);
@@ -449,11 +461,6 @@ function maybe_toggle_alphabetize_button($sortable_element: JQuery): void {
     const alphabetized_array = get_alphabetized_profile_field_choices(sortable_instance);
     const is_alphabetized = _.isEqual(choices_array, alphabetized_array);
 
-    // This is safe because only one profile field modal is open at a time ever.
-    const $alphabetize_choices_button = $(
-        ".profile-field-choices-wrapper > button.alphabetize-choices-button",
-    );
-
     if (!is_alphabetized) {
         $alphabetize_choices_button.prop("disabled", false);
         return;
@@ -586,7 +593,7 @@ function open_edit_form_modal(this: HTMLElement): void {
             "button.delete-choice",
             function (this: HTMLElement) {
                 delete_choice_row_for_edit(this, $profile_field_form, field);
-                maybe_toggle_alphabetize_button($edit_profile_field_choices_container);
+                maybe_toggle_alphabetize_button($edit_profile_field_choices_container, false);
             },
         );
         $profile_field_form.on(
@@ -855,7 +862,7 @@ function set_up_select_field(): void {
     );
     $profile_field_choices.on("click", "button.delete-choice", function (this: HTMLElement) {
         delete_choice_row(this);
-        maybe_toggle_alphabetize_button($profile_field_choices);
+        maybe_toggle_alphabetize_button($profile_field_choices, false);
     });
     $("#profile_field_choices_row").on(
         "click",
```

sanchi-t reviewed (changes requested):
I left some comments. Also, during manual testing, choices with spaces as prefixes cause entries like " z" to appear before "a" due to the leading space during sorting. We should probably discuss with CZO what the expected behavior should be. One consideration is whether we should strip spaces before sorting.

sanchi-t commented:
@tnmkr This PR ready for next round of reviews.

tnmkr commented:
> I left some comments. Also, during manual testing, choices with spaces as prefixes cause entries like " z" to appear before "a" due to the leading space during sorting. We should probably discuss with CZO what the expected behavior should be. One consideration is whether we should strip spaces before sorting.

I think we want to alphabetize without whitespace, so I've added a follow up commit which should be squashed when merging. But thanks for flagging, highlighting to get opinions from others.

tnmkr commented:
@m-e-l-u-h-a-n Marking for review from you.

@zulipbot remove "buddy review"
@zulipbot add "mentor review"

m-e-l-u-h-a-n commented on web/src/settings_profile_fields.ts:
```diff
@@ -398,6 +399,76 @@ function disable_submit_btn_if_no_property_changed(
+
+    for (const [choice_index, choice_id] of empty_choices_array) {
+        choices_array.splice(choice_index, 0, ["", choice_id]);
+    }
```
Deciding positions for empty fields after sorting.
I think this deserves a discussion with Alya and Tim, so flagging it here. Although this approach of keeping them at original position looks good to me otherwise all these empty fields would align at the beginning.

tnmkr replied:
Following #design conversation, we now move empty fields to bottom when sorting.

m-e-l-u-h-a-n reviewed (commented):
I posted a minor comment that can be discussed during maintainer review, when @timabbott or @alya review it.

Also I'm not sure if we should keep the `Alphabetize choices` button disabled when choices are sorted as maintaining that state of button appears a bit costly and hacky. Maybe we can keep just keep the button enabled and use debouncing to avoid mutliple calls to the sorting when someone clicks the button recklessly.

Although all these thoughts can be discussed during maintainer reviews, so @tnmkr you may mark it for maintainer reviews if there aren't any other changes you are planning to make.

tnmkr commented:
I posted these concerns for feedback in [original CZO thread](https://chat.zulip.org/#narrow/stream/137-feedback/topic/Favorite.20editor/near/1836053).

@alya This needs some design decisions. Marking for maintainer review.

@zulipbot remove "mentor review"
@zulip add "maintainer review"

tnmkr commented:
@zulipbot add "maintainer review"

tnmkr commented:
@alya

alya commented:
I posted some thoughts on the #design thread.

tnmkr commented:
Updated with the following changes:
- Removed follow-up commit to trim whitespaces. (from #design)
- No longer disable the button. (from #design) 
- Move empty choices to last when sorting (from #design) 
- Use `utils.strcmp` as it is more performant and used in other places.

alya commented:
Works in my manual testing! @sahil839 are you up for doing a review pass on this one?

sahil839 commented:
Looks good to me. I am not sure about using delete button visibility to check whether to add a new row or not, but cannot think of a better solution which handles all cases, so it is probably fine for now.

sahil839 commented:
One way can be to check if any input is empty among the list of inputs, but need to check how it would behave if user empties one of the existing option input.

atriCoder commented:
> One way can be to check if any input is empty among the list of inputs, but need to check how it would behave if user empties one of the existing option input.

Hey @sahil839, if you have any suggestions on how to proceed, I'd be happy to take this PR to completion since it seems inactive for a while.

shuklamaneesh23 commented:
> One way can be to check if any input is empty among the list of inputs, but need to check how it would behave if user empties one of the existing option input.

Hey @sahil839, if you have any suggestions on how to proceed, I'd be happy to take this PR to completion since it seems inactive for a while.

sahil839 commented:
This is mostly fine, except there might be a better way to check delete button visibility as I mentioned above. Would be good to confirm with @tnmkr about his availability once before working on this, since it might be quick for him to complete it than someone else.

shuklamaneesh23 commented:
Hey @tnmkr, are you available to work on completing this PR? If not, could you share your thoughts on how you'd proceed with the solution? I'd be happy to take it from there and finish it up.

tnmkr commented:
I cannot work on this for at least another week. 

Feel free to create a PR @shuklamaneesh23 , I'd be happy to review.

shuklamaneesh23 commented:
Hello @sahil839  and @tnmkr , I have filed a PR #32660 fixing this.
Please have a look at it.
Thanks!

sahil839 commented:
Closing this as #32660 fixes this. Thanks for all your work on this @tnmkr.
